### PR TITLE
Move bitfields into the IR

### PR DESCRIPTION
**Pull request by fitzgen**
This PR moves bitfields and the computation of their allocation units into the IR. They were previously computed on-the-fly during code generation. In addition to breaking up and compartmentalizing a portion of the gargantuan `CodeGeneration` implementation for `CompInfo`, this paves the way for adding const-fn bitfield unit constructors.

Also some improvements to `tests/test-one.sh`.

Also some graphviz output improvements that got tangled up in the last commit without a good way to pull them out into their own commit (sorry).

r? @emilio

### CODE REVIEW

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -736,110 +737,341 @@ impl<'a> TryToRustTy for Vtable<'a> {
...
-                      methods: &mut Vec<ast::ImplItem>)
-                      -> Layout {
-        // NOTE: What follows is reverse-engineered from LLVM's
-        // lib/AST/RecordLayoutBuilder.cpp
-        //
-        // FIXME(emilio): There are some differences between Microsoft and the
-        // Itanium ABI, but we'll ignore those and stick to Itanium for now.
-        //
-        // Also, we need to handle packed bitfields and stuff.
-        // TODO(emilio): Take into account C++'s wide bitfields, and
```
Comment (emilio) [inline src/codegen/mod.rs]: Did these fixmes go somewhere?

File: src/ir/comp.rs [hunk 2]
```diff
@@ -1,15 +1,23 @@
 //! Compound types (unions and structs) in our intermediate representation.
 
+extern crate peeking_take_while;
```
Comment (emilio) [inline src/ir/comp.rs]: Let's move it to `lib.rs`.

File: src/ir/comp.rs [hunk 3]
```diff
@@ -98,78 +106,565 @@ impl Method {
...
+    }
+
+    /// Get the layout within which these bitfields reside.
+    pub fn layout(&self) -> Layout {
+        self.layout
+    }
+
+    /// Get the bitfields within this unit.
+    pub fn bitfields(&self) -> &[Bitfield] {
+        &self.bitfields[..]
```
Comment (emilio) [inline src/ir/comp.rs]: nit: no need for `[..]`

File: src/ir/comp.rs [hunk 4]
```diff
@@ -98,78 +106,565 @@ impl Method {
...
+    }
+}
+
 /// A struct representing a C++ field.
+#[derive(Debug)]
+pub enum Field {
+    /// A normal data member.
+    DataMember(FieldData),
+
+    /// An physical allocation unit containing many logical bitfields.
```
Comment (emilio) [inline src/ir/comp.rs]: `An physical`?

File: src/ir/comp.rs [hunk 5]
```diff
@@ -98,78 +106,565 @@ impl Method {
...
+    Bitfields(BitfieldUnit),
+}
+
+impl Field {
+    fn has_destructor(&self, ctx: &BindgenContext) -> bool {
+        match *self {
+            Field::DataMember(ref data) => ctx.resolve_type(data.ty).has_destructor(ctx),
+            Field::Bitfields(BitfieldUnit { ref bitfields, .. }) => bitfields.iter().all(|b| {
```
Comment (emilio) [inline src/ir/comp.rs]: I think you can't have any field with dtors in bitfields?

File: src/ir/comp.rs [hunk 6]
```diff
@@ -98,78 +106,565 @@ impl Method {
...
+    // NOTE: What follows is reverse-engineered from LLVM's
+    // lib/AST/RecordLayoutBuilder.cpp
+    //
+    // FIXME(emilio): There are some differences between Microsoft and the
+    // Itanium ABI, but we'll ignore those and stick to Itanium for now.
+    //
+    // Also, we need to handle packed bitfields and stuff.
+    //
+    // TODO(emilio): Take into account C++'s wide bitfields, and
+    // packing, sigh.
```
Comment (emilio) [inline src/ir/comp.rs]: Oh, here they are :)

File: src/ir/comp.rs [hunk 7]
```diff
@@ -422,7 +936,12 @@ impl CompInfo {
 
     /// Get this type's set of fields.
     pub fn fields(&self) -> &[Field] {
-        &self.fields
+        match self.fields {
+            CompFields::AfterComputingBitfieldUnits(ref fields) => &fields[..],
```
Comment (emilio) [inline src/ir/comp.rs]: nit: I think you can `s/&fields[..]/fields/`

File: src/ir/context.rs [hunk 8]
```diff
@@ -160,6 +161,10 @@ pub struct BindgenContext<'ctx> {
     /// uses. See `ir::named` for more details. Always `Some` during the codegen
     /// phase.
     used_template_parameters: Option<HashMap<ItemId, ItemSet>>,
+
+    /// The set of `TypeKind::Comp` items found during parsing that need their
+    /// bitfield allocation units computed. Drained in `compute_bitfield_units`.
+    need_bitfield_allocation: Vec<ItemId>,
```
Comment (emilio) [inline src/ir/context.rs]: So this is pretty much done so we have resolved all the types before this, right? Why is this needed? I mean, we have the bitfield offset already from clang.

If this is just for simplicity, then fine, but I'd like to understand the reasoning to defer this work.

File: src/ir/comp.rs (hunk 2 above)
Comment (fitzgen) [inline src/ir/comp.rs]: Sure thing.

File: src/ir/comp.rs (hunk 3 above)
Comment (fitzgen) [inline src/ir/comp.rs]: I always forget when auto deref will be nice to me ;)

File: src/ir/comp.rs (hunk 4 above)
Comment (fitzgen) [inline src/ir/comp.rs]: Good catch, thanks

File: src/ir/comp.rs (hunk 5 above)
Comment (fitzgen) [inline src/ir/comp.rs]: Yes, that's correct, I even have a comment elsewhere to the effect for other reasons -- great catch!

File: src/ir/context.rs (hunk 8 above)
Comment (fitzgen) [inline src/ir/context.rs]: We need to delay the bitfield allocation unit computation until after we have resolved type refs because a bitfield could be of some user defined `enum` type that we haven't resolved yet.

We could potentially further slim down this work list by eagerly going `RawField` -> `Field::DataMember` for `CompInfo`s that don't have *any* bitfields, but I don't think it will have a large impact on perfomance.

**Comment by fitzgen** [comment]
@bors-servo r=emilio

Fixed up review comments and squashed them in. Thanks for the review :)
